exporter/awskinesisexporter/internal/batch/encode_marshaler.go

```diff
@@ -0,0 +1,157 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package batch
+
+import (
+	"errors"
+
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/model/otlp"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.uber.org/multierr"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awskinesisexporter/internal/key"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/zipkin/zipkinv2"
+)
+
+type batchMarshaller struct {
+	batchOptions []Option
+	partitioner  key.Partition
+
+	logsMarshaller    pdata.LogsMarshaler
+	tracesMarshaller  pdata.TracesMarshaler
+	metricsMarshaller pdata.MetricsMarshaler
+}
+
+var _ Encoder = (*batchMarshaller)(nil)
+
+func NewEncoder(named string, batchOptions ...Option) (Encoder, error) {
+	bm := &batchMarshaller{
+		batchOptions:      batchOptions,
+		partitioner:       key.Randomized,
+		logsMarshaller:    unsupported{},
+		tracesMarshaller:  unsupported{},
+		metricsMarshaller: unsupported{},
+	}
+	switch named {
+	case "zipkin-proto", "zipkin_proto", "zipkin/proto":
+		bm.tracesMarshaller = zipkinv2.NewProtobufTracesMarshaler()
+	case "zipkin-json", "zipkin_json", "zipkin/json":
+		bm.tracesMarshaller = zipkinv2.NewJSONTracesMarshaler()
+	case "otlp", "otlp-proto", "otlp_proto", "otlp/proto":
+		bm.logsMarshaller = otlp.NewProtobufLogsMarshaler()
+		bm.metricsMarshaller = otlp.NewProtobufMetricsMarshaler()
+		bm.tracesMarshaller = otlp.NewProtobufTracesMarshaler()
+	case "otlp-json", "otlp_json", "otlp/json":
+		bm.logsMarshaller = otlp.NewJSONLogsMarshaler()
+		bm.metricsMarshaller = otlp.NewJSONMetricsMarshaler()
+		bm.tracesMarshaller = otlp.NewJSONTracesMarshaler()
+	case "jaeger", "jaeger-proto", "jaeger/proto":
+		// Jaeger encoding is a special case
+		// since the internal libraries offer no means of pdata.TraceMarshaller.
+		// In order to preserve historical behavior, a custom type
+		// is used until it can be replaced.
+		return &jaegerEncoder{
+			batchOptions: batchOptions,
+		}, nil
+	default:
+		return nil, ErrUnknownExportEncoder
+	}
+	return bm, nil
+}
+
+func (bm *batchMarshaller) Logs(ld pdata.Logs) (*Batch, error) {
+	bt := New(bm.batchOptions...)
+
+	export := pdata.NewLogs()
+	export.ResourceLogs().AppendEmpty()
+
+	var errs error
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		line := ld.ResourceLogs().At(i)
+		line.CopyTo(export.ResourceLogs().At(0))
```

Comment:
You mentioned this copying being to keep message size down. I'm still a bit confused why it's better to do this copying line-by-line approach here instead of having `MarshalLogs` iterate line-by-line generating messages

Reply:
I went for this approach since it allowed for me to reuse existing methods used by exporters (kafka in particular) but there was no means of taking `resource.At(x)` and marshalling that from what I could tell.

Reply:
Ok - can you add code comments about this?

Reply:
Can do, give me a moment and I shall make it happen.